internal/ent/hooks/event.go

```diff
@@ -0,0 +1,249 @@
+			event := soiree.NewBaseEvent(fmt.Sprintf("%s.%s", typ, op), mutation)
+			event.Properties().Set("ID", other.ID)
+
+			for _, field := range fields {
+				value, exists := mutation.Field(field)
+				if exists {
+					event.Properties().Set(field, value)
+				}
+			}
+
+			event.SetContext(context.WithoutCancel(ctx))
+			event.SetClient(pool.GetClient())
+			pool.Emit(event.Topic(), event)
+
+			return retVal, err
+		})
+	}
+}
+
+// RegisterGlobalHooks registers global hooks for the ent client with the initialized event pool
+func RegisterGlobalHooks(client *entgen.Client, pool *soiree.EventPool) {
+	client.Use(EmitEventHook(pool))
+}
+
+// InitEventPool initializes an event pool with a client and a error handler
+func InitEventPool(client interface{}) *soiree.EventPool {
+	return soiree.NewEventPool(soiree.WithPool(soiree.NewPondPool(soiree.WithMaxWorkers(100), soiree.WithName("ent_event_pool"))), soiree.WithErrorHandler(func(event soiree.Event, err error) error { // nolint: mnd
+		log.Printf("Error encountered during event '%s': %v, with payload: %v", event.Topic(), err, event.Payload())
+		return nil
+	}), soiree.WithClient(client))
+}
+
+// TODO: template out listeners for all mutation types and create hooks that allow functions to be easily called from within the codebase to perform actions on those events
+// RegisterListeners registers listeners for events on the event pool and is registered with the dbclient
+func RegisterListeners(pool *soiree.EventPool) {
+	_, err := pool.On("OrganizationSetting.OpCreate", handleCustomerCreate)
+	if err != nil {
+		log.Error().Err(ErrFailedToRegisterListener)
+	}
+
+	for _, event := range []string{"OrganizationSetting.OpUpdate", "OrganizationSetting.OpUpdateOne"} {
+		_, err = pool.On(event, handleCustomerCreate)
+		if err != nil {
+			log.Error().Err(ErrFailedToRegisterListener)
+		}
+	}
+}
+
+// handleCustomerCreate handles the creation of a customer in Stripe
+func handleCustomerCreate(event soiree.Event) error {
+	props := event.Properties()
+	// TODO: add funcs for default unmarshalling of props and send all props to stripe
+	orgsettingID, exists := props["ID"]
+	if !exists {
+		log.Info().Msg("organizationSetting ID not found in event properties")
+		return nil
+	}
+
+	// if stripe ID is being added to the properties, we enumerate through any other fields in the update and ensure they match
+	stripeID, exists := props["stripe_id"]
+	if exists && stripeID != "" {
+		stripes := stripeID.(string)
+
+		stripecustomer, err := event.Client().(*entgen.Client).EntitlementManager.Client.Customers.Get(stripes, nil)
+		if err != nil {
+			log.Err(err).Msg("Failed to retrieve Stripe customer by ID")
+			return err
+		}
+
+		billingEmail, exists := props["billing_email"]
+		if exists && billingEmail != "" {
+			email := billingEmail.(string)
+			if stripecustomer.Email != email {
+				_, err := event.Client().(*entgen.Client).EntitlementManager.Client.Customers.Update(stripes, &stripe.CustomerParams{
```

Comment:
this `Update` call is happening three times in this function, it would be more efficient to do this in a single update vs. individually per field

Reply:
do you have a suggestion on how to do that given i need to check to see if each field exists?